node_modules/col-categories/lib/api.js

```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright 2015 UC Berkeley (UCB) Licensed under the
+ * Educational Community License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may
+ * obtain a copy of the License at
+ *
+ *     http://opensource.org/licenses/ECL-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an "AS IS"
+ * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+var Joi = require('joi');
+
+var DB = require('col-core/lib/db');
+var log = require('col-core/lib/logger')('col-categories');
+
+/**
+ * Get a category
+ *
+ * @param  {Context}        ctx                         Standard context containing the current user and the current course
+ * @param  {Number}         id                          The id of the category that should be retrieved
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Category}       callback.category           The retrieved category
+ * @api private
+ */
+var getCategory = module.exports.getCategory = function(ctx, id, callback) {
+  var options = {
+    'where': {
+      'course_id': ctx.course.id,
+      'id': id
+    }
+  };
+
+  DB.Category.find(options).complete(function(err, category) {
+    if (err) {
+      log.error({'err': err, 'id': id}, 'Failed to retrieve a category');
+      return callback({'code': 500, 'msg': err.message});
+    } else if (!category) {
+      log.debug({'err': err, 'id': id}, 'The category could not be found');
+      return callback({'code': 404, 'msg': 'The category could not be found'});
+    }
+
+    return callback(null, category);
+  });
+};
+
+/**
+ * Get the categories for the current course
+ *
+ * @param  {Context}        ctx                         Standard context containing the current user and the current course
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Category[]}     callback.categories         The categories in the current course
+ */
+var getCategories = module.exports.getCategories = function(ctx, callback) {
+  // Get the categories from the DB. As we need to include the number of assets per category,
+  // a raw query is used
+  var sqlQuery = 'SELECT category.*, (SELECT COUNT(*)::int FROM assets_categories WHERE category_id = category.id) AS asset_count';
```

Comment:
Although I doubt it'll ever matter and it depends on things like indexes and the set of data in the tables, doing a subselect to get the count might be suboptimal.

I would prefer something such as:
```
SELECT category.*, count(assets_categories.asset_id) AS asset_count
FROM categories AS category 
JOIN assets_categories ON assets_categories.category_id = category.id
WHERE category.course_id = 1
GROUP BY category.id;
```